### PR TITLE
Adding a test for FD.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: backends/v3/src/queue.rs [hunk 1]
```diff
@@ -364,7 +364,7 @@ impl State {
                         // Add it back to the front
                         tracing::debug!("Over budget: not enough free blocks");
                         self.entries.push_front((id, entry));
-                        break;
+                        continue;
```
Comment (Narsil) [inline backends/v3/src/queue.rs:367]: Important so we either allocate the rest or push them back into entries (otherwise they get discarded without being run)

File: backends/v3/src/queue.rs [hunk 2]
```diff
@@ -436,6 +436,12 @@ impl State {
             batch_entries.insert(id, entry);
         }
 
+        // Empty batch
+        if batch_requests.is_empty() {
+            tracing::debug!("Filterered out all entries");
+            return None;
```
Comment (Narsil) [inline backends/v3/src/queue.rs:442]: This is important because sending empty prefill is not really nice.

File: backends/v3/src/radix.rs [hunk 3]
```diff
@@ -1,10 +1,17 @@
 use crate::block_allocator::{Allocator, BlockAllocation};
 use slotmap::{DefaultKey, SlotMap};
+use std::hash::{Hash, Hasher};
 use std::{
     collections::{BTreeSet, HashMap},
     sync::Arc,
 };
 
+fn hash_(slice: &[u32]) -> u32 {
```
Comment (Narsil) [inline backends/v3/src/radix.rs:9]: The crux of the change. Instead of matching on a single token (input_ids) we match on the entire block content.

File: backends/v3/src/radix.rs [hunk 4]
```diff
@@ -1,10 +1,17 @@
 use crate::block_allocator::{Allocator, BlockAllocation};
 use slotmap::{DefaultKey, SlotMap};
+use std::hash::{Hash, Hasher};
 use std::{
     collections::{BTreeSet, HashMap},
     sync::Arc,
 };
 
+fn hash_(slice: &[u32]) -> u64 {
```
Comment (Narsil) [inline backends/v3/src/radix.rs:9]: The crux of the change. Instead of matching on a single token (input_ids) we match on the entire block content.

Comment (danieldk) [inline backends/v3/src/radix.rs:9]: For slice length 1, we can return `slice[0] as u64`. Maybe also an assertion that the slice length is at least 1?

Comment (Narsil) [reply backends/v3/src/radix.rs]: Done

File: backends/v3/src/queue.rs (hunk 1 above)
Comment (OlivierDehaene) [reply backends/v3/src/queue.rs:367]: I missed the PR where this was moved.
Since we check the prefill token budget early, if all entries in `batch`end up being in the radix tree, can't we end up running at very low prefill utilization then? 
If I understand all the components correctly, the block allocator is the only one to truly know the length in tokens that each request will have.

File: flake.lock [hunk 5]
```diff
@@ -978,11 +978,11 @@
...
-        "narHash": "sha256-nJHA1SvIQbXySpL2ueNbzQOhnkQASa5tOLz/kdW0PWA=",
+        "lastModified": 1726229792,
+        "narHash": "sha256-9xsLmjc9nr7a4PTddKv2DOi82ompTtJNyjO6R67y5tE=",
         "owner": "danieldk",
         "repo": "tgi-nix",
-        "rev": "d40f3c22e9bcc5e16c94d4605cf6a7d74dd07f46",
+        "rev": "1a902f4818e94c3f8d95f6000db17bc3fadd0ce7",
         "type": "github"
       },
       "original": {
```
Comment (danieldk) [inline flake.lock:988]: Fine with updating, but the lockfile changes seem unrelated?

**Review by danieldk — APPROVED** [review]
The radix tree changes look good to me!

File: flake.lock (hunk 5 above)
Comment (Narsil) [reply flake.lock:988]: I just did `nix flake update` at some point.

Any issue with that?
